src/main/java/org/wildfly/security/sasl/util/AuthenticationTimeoutSaslServerFactory.java

```diff
@@ -0,0 +1,154 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.SaslServer;
+import javax.security.sasl.SaslServerFactory;
+import javax.security.sasl.SaslException;
+
+import org.wildfly.security.sasl.WildFlySasl;
+
+/**
+ * A {@link SaslServerFactory} which adds authentication timeout functionality to a delegate {@code SaslServerFactory}.
+ * <p>
+ * This {@link SaslServerFactory} must be outside of the {@link AuthenticationCompleteCallbackSaslServerFactory} in the
+ * chain of SASL server factories.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+public final class AuthenticationTimeoutSaslServerFactory extends AbstractDelegatingSaslServerFactory {
+
+    /**
+     * The default amount of time, in seconds, after which an authentication attempt should time out.
+     */
+    public static final long DEFAULT_TIMEOUT = 300;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param delegate the delegate {@code SaslServerFactory}
+     */
+    public AuthenticationTimeoutSaslServerFactory(final SaslServerFactory delegate) {
+        super(delegate);
+    }
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName, final Map<String, ?> props, final CallbackHandler cbh) throws SaslException {
+        final long timeout;
+        if (props.containsKey(WildFlySasl.AUTHENTICATION_TIMEOUT)) {
+            timeout = Long.parseLong((String) props.get(WildFlySasl.AUTHENTICATION_TIMEOUT));
+        } else {
+            timeout = DEFAULT_TIMEOUT;
+        }
+        final NameAssignedCallbackHandler nameAssignedCallbackHandler = new NameAssignedCallbackHandler(cbh);
+        final SaslServer delegateSaslServer = delegate.createSaslServer(mechanism, protocol, serverName, props, nameAssignedCallbackHandler);
+        return delegateSaslServer == null ? null : new AbstractDelegatingSaslServer(delegateSaslServer) {
+            private final AtomicBoolean complete = new AtomicBoolean();
+            private final AtomicBoolean nameAssigned = new AtomicBoolean();
+            private volatile ScheduledThreadPoolExecutor executor;
+            private volatile ScheduledFuture<Void> timeoutTask;
+
+            @Override
+            public byte[] evaluateResponse(final byte[] response) throws SaslException {
+                try {
+                    final byte[] challenge = delegate.evaluateResponse(response);
+                    if (nameAssignedCallbackHandler.isNameAssigned() && nameAssigned.compareAndSet(false, true)) {
+                        // Schedule a task to terminate the authentication attempt if it takes too long
+                        executor = createExecutor();
```

Comment:
Ideally we'd use a shared executor here, which might be configured in the SecurityDomain (and in WildFly, injected from a single shared source).  This can be done as an enhancement later, but it should be an urgent one as creating a thread for every concurrent authentication could be severely problematic for most real-world cases.